### PR TITLE
StreamField migration helpers

### PR DESCRIPTION
**Pull request by gasman**
Helper functions for writing StreamField data migrations, imported from https://github.com/sandilsranasinghe/wagtail-streamfield-migration-toolkit. Migration operation code has been moved over, and now exists under `wagtail.blocks.migrations`; anything relating to auto-detection and management commands is left out for now.

Currently blocked by wagtail-factories compatibility issue with Wagtail 4.1: https://github.com/wagtail/wagtail-factories/issues/65 - there's an open question there about whether wagtail-factories should be a testing dependency of Wagtail, or should be merged into Wagtail itself, or these tests should be rewritten to not depend on it.

TODO:

- [x] version-bump wagtail-factories when a compatible version is available (or address in some other way as above)
- [x] documentation (can probably be largely copied across from wagtail-streamfield-migration-toolkit)
- [ ] more testing on real-world sites

### CODE REVIEW

**Comment by jams2** [comment]
PR with documentation opened against @gasman's branch: https://github.com/gasman/wagtail/pull/7

**Comment by jams2** [comment]
[wagtail-factories 4.0.0 release](https://github.com/wagtail/wagtail-factories/releases/tag/4.0.0)

Opened https://github.com/gasman/wagtail/pull/8 to bump the requirement.

File: docs/topics/streamfield.md [hunk 1]
```diff
@@ -8,6 +8,8 @@ For further background on StreamField, and why you would use it instead of a ric
 
 StreamField also offers a rich API to define your own block types, ranging from simple collections of sub-blocks (such as a 'person' block consisting of first name, surname and photograph) to completely custom components with their own editing interface. Within the database, the StreamField content is stored as JSON, ensuring that the full informational content of the field is preserved, rather than just an HTML representation of it.
 
+For more information on how `StreamField` interacts with Django's migration system, and a guide to migrating rich text to `StreamFields`, see the [StreamField migrations page](streamfield_migrations).
```
Comment (gasman) [inline docs/topics/streamfield.md:11]: This should probably go at the end of the page. We want to present information in the order that users need to learn it - at this point in the document, when they're only just learning about what a StreamField is, mentioning "you might to know how to migrate them" is a dangling thread they don't need.

File: docs/advanced_topics/streamfield_migrations.md [hunk 2]
```diff
@@ -0,0 +1,566 @@
...
+        migrations.RunPython(
+            convert_to_streamfield,
+            convert_to_richtext,
+        ),
+    ]
+```
+
+## StreamField data migrations
+
+As of version 4.2, Wagtail includes code from the [wagtail-streamfield-migration-toolkit](https://github.com/wagtail/wagtail-streamfield-migration-toolkit) package, which was created as part of Wagtail's participation in Google Summer of Code, 2022. The migration utilities from this package are exposed through:
```
Comment (gasman) [inline docs/advanced_topics/streamfield_migrations.md:182]: Other than a `versionadded` note (which we can systematically search for and drop at the appropriate time), we should avoid talking about how and when a feature was introduced - that's not relevant information to someone who just wants to use it. Anyone who is interested in changes to Wagtail over time can look at the release notes.

**Pull request merged**